docs/source/integrations/building-integrations.md

```diff
@@ -0,0 +1,71 @@
+---
+title: Building Integrations for Apollo Server
+description: ""
+---
+
+One of the driving forces behind Apollo Server 4 is the creation of a stable, well-defined API for processing HTTP requests and responses. Apollo Server 4's API enables external collaborators (like you) to build integrations with Apollo Server in their web framework of choice.
+
+The primary responsibility of an Apollo Server integration is to translate requests and responses between a web framework's native format to the format that `ApolloServer` uses.
+## Integration patterns
+
+<!-- TODO: fix link to point to main once version-4 is merged  -->
+> See the [`expressMiddleware` function](https://github.com/apollographql/apollo-server/blob/36482f5eb56a0421c1eb47e3ebf0e60e033573ab/packages/server/src/express/index.ts) for an example of integrating [Express](https://github.com/expressjs/express) with Apollo Server.
+
+Your integration should expect to intake an `ApolloServer` instance _after_ that instance has called the `server.start()` method:
+
+```ts
+// Create a new instance of ApolloServer
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+});
+
+// The async start method instructs Apollo Server to
+// prepare to handle incoming operations.
+await server.start();
+```
+
+You can then pass the "started" `ApolloServer` instance into your framework-specific middleware. For example, the `expressMiddleware` function accepts an `ApolloServer` instance, like so:
+
+```ts
+// 1. Create our ApolloServer instance and start it
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+});
+await server.start();
+
+// 2. Create our Express server
+const app = express();
+const httpServer = http.createServer(app);
+
+// 3. Set up our middleware for our Express server.
+// The expressMiddleware function accepts an ApolloServer instance,
+// attaching Apollo Server to our Express server.
+app.use(expressMiddleware(server));
+```
+
+After setting up the framework-specific middleware, you can start your chosen framework's server:
+
+```ts
+// 1. Create our ApolloServer instance and start it
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+});
+await server.start();
+
+// 2. Create our framework specific server
+const app = express();
+const httpServer = http.createServer(app);
+
+// 3. Set up our middleware for our framework specific server.
+app.use(cors(), bodyParser.json(), expressMiddleware(server));
+
+// 4. Start the framework specific server
+await new Promise<void>((resolve) => httpServer.listen({ port: 4000 }, resolve));
+console.log(`🚀 Server ready at http://localhost:4000`);
+```
```

Comment:
I suspect you've taken inspiration from our express tests (or maybe the drain plugin docs), which are both great places to look. I'm going to elaborate a bit, hopefully not too much but let me know if too little!

Generally (in typical express server use cases), express handles `listen`ing. This means it's responsible for receiving the request, passing it along to its middleswares, and responding. In this case, there's really no need for an additional `httpServer` (since they do the same job!).

ApolloServer has a problem: when it's told to stop, it can't really tell express to stop accepting requests. So you might end up with the case where AS is stopping but express has accepted requests and forwarded them along.

The `httpServer` is our way to take control of how this works. If `httpServer` is responsible for listening, it looks like:
request -> httpServer -> express -> middlewares

ApolloServer knows how to stop an `httpServer` via `ApolloServerPluginDrainHttpServer`, which tells the `httpServer` to stop accepting incoming requests while AS finishes the ones currently in flight.

All of this is to say:
* If we want to educate on `ApolloServerPluginDrainHttpServer` here, then we should continue using `httpServer` (but add the plugin usage).
* If we want to keep this example simple, we should get rid of `httpServer` and just use express via `app.listen()`.

More complete explanation here: https://www.apollographql.com/docs/apollo-server/api/plugin/drain-http-server/

Also I think these 2 separate code sections are mostly duplicated (minus step 4). Would it be better just to combine them?